src/hotspot/cpu/aarch64/aarch64.ad

```diff
@@ -6448,6 +6448,19 @@ instruct loadNKlass(iRegNNoSp dst, memory mem)
   ins_pipe(iload_reg_mem);
 %}
 
+instruct loadNKlassCompactHeaders(iRegNNoSp dst, memory mem)
+%{
+  match(Set dst (LoadNKlass mem));
+  predicate(!needs_acquiring_load(n) && UseCompactObjectHeaders);
+
+  ins_cost(4 * INSN_COST);
+  format %{ "ldrw  $dst, $mem\t# compressed class ptr" %}
+  ins_encode %{
+    __ load_nklass_compact_c2($dst$$Register, $mem$$base$$Register, $mem$$index$$Register, $mem$$scale, $mem$$disp);
```

Comment:
I wonder if something along the line of this is required here.
```suggestion
    Address addr = mem2address($mem->opcode(), $mem$$base$$Register, $mem$$index, $mem$$scale, $mem$$disp);
    __ load_nklass_compact_c2($dst$$Register, __ adjust_compact_object_header_address_c2(addr, rscratch1));
```
With `adjust_compact_object_header_address_c2` being:
```C++
Address C2_MacroAssembler::adjust_compact_object_header_address_c2(Address addr, Register tmp) {
  // The incoming address is pointing into obj-start + klass_offset_in_bytes. We need to extract
  // obj-start, so that we can load from the object's mark-word instead. Usually the address
  // comes as obj-start in addr.base() and klass_offset_in_bytes in addr.offset().
  if (addr.getMode() != Address::base_plus_offset) {
    lea(tmp, addr);
    addr = Address(tmp, -oopDesc::klass_offset_in_bytes());
  } else {
    addr = Address(addr.base(), addr.offset() - oopDesc::klass_offset_in_bytes());
  }
  return legitimize_address(addr, 8, tmp);
}
```

Maybe it is the case that we never get the case where `$mem->opcode()` is not `lsl` variant, nor that the offset is to far away for an immediate fixed by `legitimize_address`. But it seems like this would at least make those cases correct, while avoiding the `lea` in the common case. 

Maybe someone with better experience in aarch64 macroassembler+ad files and C2 can give an opinion.